Tests/NIOTests/ByteBufferTest.swift

```diff
@@ -1524,30 +1524,144 @@ class ByteBufferTest: XCTestCase {
+
+        // Replace with longer range
+        view[5..<8] = oneToNineView[5..<9]
+        XCTAssertTrue(view.elementsEqual(oneToNine))
+    }
+
+    func testBufferViewReplaceSubrangeWithEqualLengthBytes() throws {
+        self.buf.writeBytes([0x0, 0x1, 0x2, 0x3, 0x4])
+
+        var view = ByteBufferView(takingOwnershipOf: self.buf)
+        XCTAssertEqual(view.count, self.buf.readableBytes)
+
+        view.replaceSubrange(view.indices.suffix(3), with: [0xd, 0xe, 0xf])
+
+        var modifiedBuf = ByteBuffer(takingOwnershipOf: view)
+        XCTAssertEqual(self.buf.readerIndex, modifiedBuf.readerIndex)
+        XCTAssertEqual(self.buf.writerIndex, modifiedBuf.writerIndex)
+        XCTAssertEqual([0x0, 0x1, 0xd, 0xe, 0xf], modifiedBuf.readBytes(length: modifiedBuf.readableBytes)!)
+    }
+
+    func testBufferViewReplaceSubrangeWithFewerBytes() throws {
+        self.buf.writeBytes([0x0, 0x1, 0x2, 0x3, 0x4])
+
+        var view = ByteBufferView(takingOwnershipOf: self.buf)
+        view.replaceSubrange(view.indices.suffix(3), with: [0xd])
+
+        var modifiedBuf = ByteBuffer(view)
+        XCTAssertEqual(self.buf.readerIndex, modifiedBuf.readerIndex)
+        XCTAssertEqual(self.buf.writerIndex - 2, modifiedBuf.writerIndex)
+        XCTAssertEqual([0x0, 0x1, 0xd], modifiedBuf.readBytes(length: modifiedBuf.readableBytes)!)
+    }
+
+    func testBufferViewReplaceSubrangeWithMoreBytes() throws {
+        self.buf.writeBytes([0x0, 0x1, 0x2, 0x3])
+
+        var view = ByteBufferView(takingOwnershipOf: self.buf)
+        XCTAssertTrue(view.elementsEqual([0x0, 0x1, 0x2, 0x3]))
+
+        view.replaceSubrange(view.indices.suffix(1), with: [0xa, 0xb])
+        XCTAssertTrue(view.elementsEqual([0x0, 0x1, 0x2, 0xa, 0xb]))
+
+        var modifiedBuf = ByteBuffer(takingOwnershipOf: view)
+        XCTAssertEqual(self.buf.readerIndex, modifiedBuf.readerIndex)
+        XCTAssertEqual(self.buf.writerIndex + 1, modifiedBuf.writerIndex)
+        XCTAssertEqual([0x0, 0x1, 0x2, 0xa, 0xb], modifiedBuf.readBytes(length: modifiedBuf.readableBytes)!)
+    }
+
+    func testBufferViewEmpty() throws {
+        self.buf.writeBytes([0, 1, 2])
+
+        var view = ByteBufferView()
+        view[0..<0] = self.buf.view
+        XCTAssertEqual(view.indices, 0..<3)
+
+        self.buf = ByteBuffer(takingOwnershipOf: view)
+        XCTAssertEqual(self.buf.readerIndex, 0)
+        XCTAssertEqual(self.buf.writerIndex, 3)
+
+        let anotherBuf = self.buf!
+        XCTAssertEqual([0, 1, 2], self.buf.readBytes(length: self.buf.readableBytes))
+
+        var anotherView = anotherBuf.view
+        anotherView.replaceSubrange(0..<3, with: [])
+        XCTAssertTrue(anotherView.isEmpty)
+
+        self.buf = ByteBuffer(takingOwnershipOf: anotherView)
+        XCTAssertEqual(self.buf.readerIndex, 0)
+        XCTAssertEqual(self.buf.writerIndex, 0)
+        XCTAssertEqual([], self.buf.readBytes(length: self.buf.readableBytes))
+    }
```

Comment:
we need some tests that prove that you can hold a mutable `ByteBufferView` in a way that it doesn't actually trigger a CoW.

Reply:
There's an allocation test at the top of this PR, unfortunately Swift can't currently doesn't avoid the CoW but as of https://github.com/apple/swift/commit/6acf8db2468ee1aaed346f1fc38a5179841722b4 this has been fixed (confirmed by re-running the alloc test)